Add ExecutionPolicy to SubtypeEffectHandler

juanmrivero opened the pull request:
This PR introduces the parameter `executionPolicy: ExecutionPolicy` too all builder functions in `CoroutinesSubtypeEffectHandlerBuilder`. 

`ExecutionPolicy` gives users simple policies to control concurrency while processing effects for each effect type in the builder. The current implemented policies are: 
- `RunSequentially`: all effects of the same type wait for the previous one to finish before executing.
- `RunConcurrently`: all effects of the same type are executed immediately and concurrently.
- `CancelPrevious`: a new effect of the same type start  executing immediately and cancel the execution of any previously running effect.

Here is an example of usage: 
```kotlin
    Mobius.loop(
            Update(...),
            MobiusCoroutines.subtypeEffectHandler<Effect, Event>()
              .addAction<Effect.Log>(executionPolicy = RunConcurrently) { ... }
              .addConsumer<Effect.UpdateSearchQuery>(executionPolicy = CancelPrevious){ effect -> ... }
              .addFunction<Effect.AddToBag>(executionPolicy = RunSequentially) { effect -> event }
             .build()
    )
``` 

By default all functions use `RunSequentially`. This was selected to avoid unintended cancellations or concurrent problems while processing an effect.

To support this feature the signature of the method `addEffectHandler(kClass: KClass<out F>, function: suspend (F) -> Flow<E>)` was changed to `addEffectHandler(kClass: KClass<out F>, effectHandler: EffectHandler<F, E>)`. **This is non backwards compatible**. I'm assuming this is ok, since the method `addFlowProducer`, or `addFlow` (introduced in this PR) are analogous and easier to use.

Additionally the function `addEffectHandler(kClass: KClass<out F>, effectHandler: EffectHandler<F, E>)` allows the user to implement its own concurrency policy by handling the effect & event channels directly in case it needs more fine grained control.

---

This PR also introduces an utility function `MobiusCoroutines.effectHandler( coroutineContext: CoroutineContext, onEffect: suspend (effect: F, eventConsumer: Consumer<E>) -> Unit)` which creates a simple effect handler that will run all effects concurrently.

 Here is an example of usage: 
```kotlin
      Mobius.loop(
            Update(...),
            MobiusCoroutines.effectHandler<Effect, Event>(){ effect, eventConsumer ->
                when(effect) {
                    is Effect.Log -> { ... }
                    is Effect.AddToBag -> {
                        ...
                        eventConsumer.accept(Event.AddToBagSuccess)
                    }
                }
            }
    )
```

pettermahlen commented on mobius-coroutines/src/test/java/com/spotify/mobius/coroutines/CoroutinesSubtypeEffectHandlerBuilderTest.kt:
```diff
@@ -41,7 +38,7 @@ class CoroutinesSubtypeEffectHandlerBuilderTest {
                 val effectHandler = subtypeEffectHandler<Effect, Event>()
 
-                effectHandler.build(coroutineContext)
+                effectHandler.build(UnconfinedTestDispatcher(testScheduler))
```
this `UnconfinedTestDispatcher(testScheduler)` is duplicated in multiple tests. Would it be possible to move to a setup method or field initialiser?

juanmrivero replied:
All tests first create the effect handler, then it is setup and lately it is built using one of the `TestDispatchers`

Since I'm using the functions `advanceUntilIdle()` I need a reference to `testScheduler` from the `TestScope` in the function `runTest`

I tried a couple of options like: 
```kotlin
    // Only build
    private fun CoroutinesSubtypeEffectHandlerBuilder<Effect, Event>.buildUnconfined(testScheduler: TestCoroutineScheduler)
        = this.build(UnconfinedTestDispatcher(testScheduler))
    
    //Build and connect    
    private fun CoroutinesSubtypeEffectHandlerBuilder<Effect, Event>.buildUnconfined(testScheduler: TestCoroutineScheduler, consumer: Consumer<Event> = Consumer {})
        = this.build(UnconfinedTestDispatcher(testScheduler)).connect(consumer)
```

But then the code: 
```kotlin
        effectHandler.build(UnconfinedTestDispatcher(testScheduler))
            .connect { }
            .accept(Effect.Simple)
```
Only gets abbreviated by:
```kotlin
        effectHandler.buildUnconfined(testScheduler))
            .accept(Effect.Simple)
```

without improving readability but obscuring a little bit what is actually happening. 

Do you have any specific function in mind to try ?

pettermahlen commented on mobius-coroutines/src/main/java/com/spotify/mobius/coroutines/CoroutinesSubtypeEffectHandlerBuilder.kt:
```diff
@@ -23,68 +30,152 @@ import kotlin.reflect.KClass
         }
     }
+
+    sealed interface CancellationPolicy {
```
Wouldn't `ExecutionPolicy` be a more appropriate name given the available implementations?

juanmrivero replied:
Great suggestion ! I changed the name

pettermahlen commented on mobius-coroutines/src/main/java/com/spotify/mobius/coroutines/CoroutinesSubtypeEffectHandlerBuilder.kt:
```diff
@@ -23,68 +30,152 @@ import kotlin.reflect.KClass
-        }
+        addEffectHandler(G::class, when (cancellationPolicy) {
+
+            RunSequentially -> { effectChannel, eventsChannel ->
+                for (effect in effectChannel) {
+                    function.invoke(effect as G, eventsChannel)
+                }
+            }
+
+            RunConcurrently -> { effectChannel, eventsChannel ->
+                coroutineScope {
+                    for (effect in effectChannel) {
+                        launch { function.invoke(effect as G, eventsChannel) }
+                    }
+                }
+            }
+
+            CancelPrevious -> { effectChannel, eventsChannel ->
+                coroutineScope {
+                    var currentJob: Job = Job()
+                    for (effect in effectChannel) {
+                        currentJob.cancel()
+                        currentJob = launch { function.invoke(effect as G, eventsChannel) }
+                    }
+                }
+            }
```
If this code would be embedded in the respective strategy objects instead of implemented here, I think that would open up for multiple/custom implementations of the policy. 

So, basically, something like `executionPolicy.create(effectChannel, eventsChannel).run(function)`, maybe?

juanmrivero replied:
Great suggestion. I changed `ExecutionPolicy` to be an `interface` instead of a `sealed interface`. That gives more flexibility when using `CoroutinesSubtypeEffectHandlerBuilder` "add..." functions.

pettermahlen reviewed (commented):
A couple of comments. Regarding the API-breaking change, I think that's fine, but that it should lead to a new major version, of course. It might be useful to consider if there are other changes that should be included. 

Also, are the japicmp API checks enabled for mobius-coroutines?

togi commented on mobius-coroutines/src/main/java/com/spotify/mobius/coroutines/CoroutinesSubtypeEffectHandlerBuilder.kt:
```diff
@@ -23,68 +30,152 @@ import kotlin.reflect.KClass
     }
 
     inline fun <reified G : F> addConsumer(
+        cancellationPolicy: CancellationPolicy = RunSequentially,
```
I thought that RunConcurrently would be the the non-breaking change? But I suppose sequential execution probably makes more sense as a default

juanmrivero replied:
Previously there were no guaranties, I picked this one as the "safe one" since it does not involves concurrency or cancellations (for the same type of event).

togi commented on mobius-coroutines/src/main/java/com/spotify/mobius/coroutines/CoroutinesSubtypeEffectHandlerBuilder.kt:
```diff
@@ -23,68 +30,152 @@ import kotlin.reflect.KClass
-        flowOf()
     }
 
     inline fun <reified G : F> addConsumer(
```
Missing documentation for the 'add*' methods, but I suppose that was already the case on master, so maybe a task for a separate PR.

juanmrivero replied:
I added the documentation in this one.

juanmrivero commented:
@pettermahlen I updated the version to `2.0.0` and updated the `baselineVersion` in `binary_compatibility.gradle` to that version. 

About if japicmp is enabled for mobius-coroutines, I understand by [this line](https://github.com/spotify/mobius/blob/8f07c969c73e37305ddf7127899dcbc5d06925e3/build.gradle#L87) that it is, since the module applies the [java-library](https://github.com/spotify/mobius/blob/8f07c969c73e37305ddf7127899dcbc5d06925e3/mobius-coroutines/build.gradle#L3) plugin
